### PR TITLE
Persist donation

### PR DESCRIPTION
**Pull request by KaiNissen**

### CODE REVIEW

File: src/Domain/DoctrineDonationRepository.php [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+	private $entityManager;
+
+	public function __construct( Connection $connection ) {
+		$factory = new FundraisingStoreFactory( $connection );
+		$this->entityManager = $factory->getEntityManager();
+	}
+
+	public function storeDonation( Donation $donation ) {
+		$this->entityManager->persist( $donation );
```
Comment (JeroenDeDauw) [inline src/Domain/DoctrineDonationRepository.php]: This is broken... nothing using it though

File: src/FunFunFactory.php [hunk 2]
```diff
@@ -576,7 +576,8 @@ public function newAddDonationUseCase(): AddDonationUseCase {
 				$this->newPaymentTypeValidator(),
 				$this->newBankDataValidator(),
 				$this->getMailValidator()
-			)
+			),
+			$this->newGeneralizedReferrer()
```
Comment (JeroenDeDauw) [inline src/FunFunFactory.php]: This is a bad class name btw: adjective + noun. This would be better: `ReffererGeneralizer`

**Comment by JeroenDeDauw** [comment]
Follow up: #183

File: src/Domain/Donation.php [hunk 3]
```diff
@@ -69,4 +109,98 @@ public function setBankData( BankData $bankData ) {
...
+
+	public function determineFullName() {
+		if ( $this->getPersonalInfo() !== null ) {
+			return $this->getPersonalInfo()->getPersonName()->getFullName();
+		}
+
+		return 'Anonym';
+	}
+
+	public function generateTransferCode() {
```
Comment (JeroenDeDauw) [inline src/Domain/Donation.php]: Where do you intend to use this? This code appears to have no cohesion with the Donation class, which suggests it does not belong here. Can't say without knowing what this is intended for though.

File: src/Domain/Donation.php [hunk 4]
```diff
@@ -13,11 +14,41 @@
...
 	private $personalInfo;
+	private $optIn;
 	private $bankData;
+	private $tracking;
+	private $source;
+	private $totalImpressionCount;
+	private $singleBannerImpressionCount;
+	private $color;
+	private $skin;
+	private $layout;
```
Comment (JeroenDeDauw) [inline src/Domain/Donation.php]: Would this all fit in a TrackingData object? It is all tracking stuff right, or is some of it not?

File: src/UseCases/AddDonation/AddDonationUseCase.php [hunk 5]
```diff
@@ -59,4 +73,68 @@ private function newBankDataFromRequest( AddDonationRequest $request ): BankData
 		return $bankData->freeze()->assertNoNullFields();
 	}
 
+	private function newDonationEntityFromDonationDomain( Donation $donation ) {
+		$dbalDonation = new \WMDE\Fundraising\Entities\Donation();
```
Comment (JeroenDeDauw) [inline src/UseCases/AddDonation/AddDonationUseCase.php]: This is misplaced. Fixed in #183

The idea is to not have anything that is a persistence implementation detail, such as the Store component or Doctine, outside of the DataAccess folder. (Well, also in the factory and tests possible, but certainly not in the Domain folder.) Not how the repository interfaces are in the Domain folder and the implementations are in DataAccess.

File: tests/System/Routes/AddDonationRouteTest.php [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+		'layout' => 'Default',
+		'color' => 'blue',
+		'skin' => 'default',
+	];
+
+	public function testGivenValidRequest_donationGetsPersisted() {
+		$donationRepository = new DonationRepositorySpy();
+
+		$client = $this->createClient( [], function( FunFunFactory $factory ) use ( $donationRepository ) {
+			$factory->setDonationRepository( $donationRepository );
```
Comment (JeroenDeDauw) [inline tests/System/Routes/AddDonationRouteTest.php]: The responsibility of the system tests is to verify the whole stack works, so it's generally better to not mock out the database. Changed in #183

File: src/DataAccess/DoctrineDonationRepository.php [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+use WMDE\Fundraising\Frontend\Domain\Model\PaymentType;
+use WMDE\Fundraising\Frontend\Domain\Repositories\DonationRepository;
+use WMDE\Fundraising\Entities\Donation as DoctrineDonation;
+use WMDE\Fundraising\Frontend\Domain\Donation;
+
+/**
+ * @license GNU GPL v2+
+ * @author Kai Nissen < kai.nissen@wikimedia.de >
+ */
+class DoctrineDonationRepository implements DonationRepository {
```
Comment (gbirke) [inline src/DataAccess/DoctrineDonationRepository.php]: Could you rename it to DbalDonationRepository? No need to encode the library name in the class.

File: src/Domain/Donation.php [hunk 8]
```diff
@@ -13,11 +14,41 @@
...
+
+	const STATUS_MODERATION = 'P';
+	const STATUS_DELETED = 'D';
+
+	private $status;
 	private $amount;
 	private $interval = 0;
 	private $paymentType;
 	private $personalInfo;
+	private $optIn;
```
Comment (gbirke) [inline src/Domain/Donation.php]: This is true when the user has opted in to receive more infos, right? or is it for othe opt-in processses? Maybe a more descriptive name?

**Comment by gbirke** [comment]
+1, apart from some name nagging that could be ignored.

File: src/DataAccess/DoctrineDonationRepository.php (hunk 7 above)
Comment (JeroenDeDauw) [inline src/DataAccess/DoctrineDonationRepository.php]: Dbal is also a library :) Actually I think our other ones should be renamed to Doctine, as they use Doctine ORM, not just DBAL

File: src/Domain/Donation.php (hunk 4 above)
Comment (JeroenDeDauw) [inline src/Domain/Donation.php]: I'm going to split this off in a follow up, hoping I'm guessing correctly

File: src/Domain/Donation.php (hunk 8 above)
Comment (JeroenDeDauw) [inline src/Domain/Donation.php]: Yeah, not very descriptive. This indeed is the `info` HTTP request parameter and `info` db field. So this should be a boolean, not a string

File: src/Domain/Donation.php (hunk 3 above)
Comment (KaiNissen) [inline src/Domain/Donation.php]: moved it to FundraisingStore (see wmde/FundraisingStore#52)

File: src/Domain/Donation.php (hunk 4 above)
Comment (KaiNissen) [inline src/Domain/Donation.php]: yes, you totally guessed right

**Pull request merged**